server/src/main/java/org/apache/iotdb/db/engine/migration/MigratingFileLogManager.java

```diff
@@ -0,0 +1,127 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.db.engine.migration;
+
+import org.apache.iotdb.db.engine.fileSystem.SystemFileFactory;
+import org.apache.iotdb.db.engine.storagegroup.TsFileResource;
+import org.apache.iotdb.tsfile.utils.FilePathUtils;
+import org.apache.iotdb.tsfile.utils.ReadWriteIOUtils;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.file.Paths;
+
+import static org.apache.iotdb.db.metadata.idtable.IDTable.config;
+import static org.apache.iotdb.db.metadata.idtable.IDTable.logger;
+
+/**
+ * To assure that migration of tsFiles is pesudo-atomic operator, MigratingFileLogManager writes
+ * files to migratingFileDir when a tsFile (and its resource/mod files) are being migrated, then
+ * deletes it after it has finished operation.
+ */
+public class MigratingFileLogManager {
+
+  private static File MIGRATING_LOG_DIR =
+      SystemFileFactory.INSTANCE.getFile(
+          Paths.get(FilePathUtils.regularizePath(config.getSystemDir()), "migration", "migrating")
+              .toString());
+
+  protected MigratingFileLogManager() {
+    if (MIGRATING_LOG_DIR == null) {
+      logger.error("MIGRATING_LOG_DIR is null");
+    }
+
+    if (!MIGRATING_LOG_DIR.exists()) {
+      if (MIGRATING_LOG_DIR.mkdirs()) {
+        logger.info("MIGRATING_LOG_DIR {} created successfully", MIGRATING_LOG_DIR);
+      } else {
+        logger.error("MIGRATING_LOG_DIR {} create error", MIGRATING_LOG_DIR);
+      }
+      return;
+    }
+
+    if (!MIGRATING_LOG_DIR.isDirectory()) {
+      logger.error("{} already exists but is not directory", MIGRATING_LOG_DIR);
+    }
+  }
+
+  // singleton
+  private static class MigratingFileLogManagerHolder {
+    private MigratingFileLogManagerHolder() {}
+
+    private static final MigratingFileLogManager INSTANCE = new MigratingFileLogManager();
+  }
+
+  public static MigratingFileLogManager getInstance() {
+    return MigratingFileLogManagerHolder.INSTANCE;
+  }
+
+  /** started migrating tsfile and its resource/mod files */
+  public boolean start(File tsfile, File targetDir) throws IOException {
+    File logFile = SystemFileFactory.INSTANCE.getFile(MIGRATING_LOG_DIR, tsfile.getName() + ".log");
+    if (!logFile.createNewFile()) {
+      return false;
+    }
+
+    FileOutputStream logFileOutput = new FileOutputStream(logFile);
+    ReadWriteIOUtils.write(tsfile.getAbsolutePath(), logFileOutput);
+    ReadWriteIOUtils.write(targetDir.getAbsolutePath(), logFileOutput);
+    logFileOutput.flush();
+    logFileOutput.close();
+
+    return true;
+  }
+
+  /** finished migrating tsfile and related files */
+  public void finish(File tsfile) {
+    File logFile = SystemFileFactory.INSTANCE.getFile(MIGRATING_LOG_DIR, tsfile.getName() + ".log");
+    if (logFile.exists()) {
+      logFile.delete();
+    }
```

Comment:
I have an doubt here, maybe you can test the performance between the following implementations:
1. Just create one log file, and log all migration information into it.
2. Create one log file for each migration task, and log its' migration information into it.
3. Create one log file for each migration file (your current implementation). I think this implementation may slow because you submit a large amount of small I/O tasks.

Reply:
I'm not sure how to sync the log for implementation 1 when there are concurrent tasks, so instead I tested:

1. Without Logs (baseline).
2. Original Implementation.
3. One log for each migration task, append to log for each tsfile migration.
4. One log for each migration task, overwrite log for each tsfile migration.

Tested on 10000 tsfiles.

| Implementation | Migration Time (ms) | Recovery Time (ms) |
| ----------- | ------- | ----------- |
| Without Logs | 400 | - |
| Original | 1400 | 2 |
| Append Log | 450 | 900 |
| Overwrite Log | 800 | 2 |

Since migrating is more common and recovery is sparse, changed to method 3 (one log per task, append log).

Reply:
Excellent work 👍

Reply:
Excellent work 👍